packages/graphql-codegen-cli/src/utils/patterns.ts

```diff
@@ -0,0 +1,308 @@
+
+function log(msg: string) {
+  // double spaces to inline the message with Listr
+  getLogger().info(`  ${msg}`);
+}
+
+type NegatedPattern = `!${string}`;
+
+/**
+ * Flatten a list of pattern sets to be a list of the original patterns that
+ * were used to create them (i.e., including both affirmative and negative).
+ *
+ * This can be used, for example, to find the "longest common prefix directory"
+ * by examining `mm.scan(pattern).base` for each `pattern`
+ */
+export const flattenPatternSets = (patternSets: PatternSet[]) => {
+  return patternSets.flatMap(patternSet => [
+    ...patternSet.watch.patterns,
+    ...patternSet.documents.patterns,
+    ...patternSet.schemas.patterns,
+  ]);
+};
+
+/**
+ * Create a rebuild trigger that follows the algorithm described here:
+ * https://github.com/dotansimha/graphql-code-generator/issues/9270#issuecomment-1496765045
+ *
+ * There is a flow chart diagram in that comment.
+ *
+ * Basically:
+ *
+ *  * "Global" patterns are defined at top level of config file, and "local"
+ *    patterns are defined for each output target
+ *  * Each pattern can have "watch", "documents", and "schemas"
+ *  * Watch patterns (global and local) always take precedence over documents and
+ *    schemas patterns, i.e. a watch negation always negates, and a watch match is
+ *    a match even if it would be negated by some pattern in documents or schemas
+ *  * The trigger returns true if any output target's local patterns result in
+ *    a match, after considering the precedence of any global and local negations
+ */
+export const makeShouldRebuild = ({
+  globalPatternSet,
+  localPatternSets,
+}: {
+  globalPatternSet: PatternSet;
+  localPatternSets: PatternSet[];
+}) => {
+  const localMatchers = localPatternSets.map(localPatternSet => {
+    return (path: string) => {
+      // Is path negated by any negating watch pattern?
+      if (
+        matchesAnyAffirmativePattern(
+          path,
+          invertNegatedPatterns([...globalPatternSet.watch.negated, ...localPatternSet.watch.negated])
+        )
+      ) {
+        // Short circut: negations in watch patterns take priority
+        return false;
+      }
+
+      // Does path match any affirmative watch pattern?
+      if (
+        matchesAnyAffirmativePattern(path, [
+          ...globalPatternSet.watch.affirmative,
+          ...localPatternSet.watch.affirmative,
+        ])
+      ) {
+        // Immediately return true: Watch pattern takes priority, even if documents or schema would negate it
+        return true;
+      }
+
+      // Does path match documents patterns (without being negated)?
+      if (pathMatches(path, [...globalPatternSet.documents.patterns, ...localPatternSet.documents.patterns])) {
+        return true;
+      }
+
+      // Does path match schemas patterns (without being negated)?
+      if (pathMatches(path, [...globalPatternSet.schemas.patterns, ...localPatternSet.schemas.patterns])) {
+        return true;
+      }
```

Comment:
For `documents` and `schemas`, we rely on `micromatch`, and we pass it a list of patterns that can include both affirmative (non-negated) and negated patterns. 

Because `micromatch` is a bit finnicky, we can't use `micromatch.isMatch`, which is why we need to make this `pathMatches` function.

Reply:
After writing test cases, it turns out this was not true - we cannot rely on micromatch - so I deleted the `patchMatches` shortcut, in favor of explicitly checking the affirmative matches, and then checking there are no negated matches. But I actually prefer it this way, because now the logic matches the flow chart exactly.

The reason for this is that **micromatch cares about order of rules**, and a negation will only override a match when it comes **after** the affirmative rule. However, we do not care about order - negations should always override affirmative matches.

See fix in b2328fc30c0b6ad519d7e0b7bb9300a57e9442f8 (but note the commit message body is inaccurate - the reason micromatch doesn't work has to do with ordering, not exact vs. glob match). 

This shows why it didn't work:

```js
// Given this config
globalPattern.watch =   '!**/exclude-doc-everywhere.graphql'
localPattern.watch =   'foo/global-beats-local/exclude-doc-everywhere.graphql'

// And this path:
const path = 'foo/global-beats-local/exclude-doc-everywhere.graphql'

// We would expect this NOT to match, because the global negation should win

// However, micromatch considers this a match:
true === require('micromatch')(
  [path],
  [
    '!**/exclude-doc-everywhere.graphql',
    'foo/global-beats-local/exclude-doc-everywhere.graphql'
     // Note: the same is true even with:
     // '**/exclude-doc-everywhere.graphql',
  ],
  { cwd: process.cwd() }
).length === 1

// But when changing the order, it does not consider it a match
false === require('micromatch')(
  [path],
  [
    'foo/global-beats-local/exclude-doc-everywhere.graphql',
     // Note: the same is true even with:
     // '**/exclude-doc-everywhere.graphql',
    '!**/exclude-doc-everywhere.graphql'
  ],
  { cwd: process.cwd() }
).length === 1
```

Reply:
> However, we do not care about order (right? or maybe we do care about order within local groups? let me know) - negations should always override affirmative matches.

I actually think we care about the order e.g. if we have the following order

```
foo/foo.graphql
!foo/foo/bar/**/*
foo/foo/bar/asdf/foo.graphql
```

This should still include `foo/foo/bar/asdf/foo.graphql` IMHO and `!foo/foo/bar/**/*` should not overwrite it. is this the current behavior?